src/strat_engine/blockdev.rs

```diff
@@ -0,0 +1,373 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::io::{Read, Write, ErrorKind, Seek, SeekFrom};
+use std::fs::{OpenOptions, read_dir};
+use std::path::{Path, PathBuf};
+use std::io;
+use std::str::{FromStr, from_utf8};
+use std::cmp::Ordering;
+
+use nix::sys::stat;
+use time::Timespec;
+use devicemapper::Device;
+use crc::crc32;
+use byteorder::{LittleEndian, ByteOrder};
+use uuid::Uuid;
+use bytesize::ByteSize;
+
+use types::{Sectors, SectorOffset};
+use engine::{EngineResult, EngineError};
+
+use consts::*;
+use super::consts::*;
+use super::util::blkdev_size;
+
+pub use super::BlockDevSave;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct MDA {
+    pub last_updated: Timespec,
+    used: u32,
+    length: u32,
+    crc: u32,
+    offset: SectorOffset, // From start of MDA, not BDA
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct BlockDev {
+    pub stratis_id: String,
+    pub dev: Device,
+    pub id: String,
+    pub path: PathBuf,
+    pub sectors: Sectors,
+    pub mdaa: MDA,
+    pub mdab: MDA,
+    mda_sectors: Sectors,
+    reserved_sectors: Sectors,
+}
+
+impl BlockDev {
+    pub fn new(stratis_id: &str,
+               path: &Path,
+               mda_size: Sectors,
+               force: bool)
+               -> EngineResult<BlockDev> {
+        let pstat = try!(stat::stat(path));
+
+        if pstat.st_mode & 0x6000 != 0x6000 {
+            return Err(EngineError::Io(io::Error::new(ErrorKind::InvalidInput,
+                                                      format!("{} is not a block device",
+                                                              path.display()))));
+        }
+
+        if *mda_size % 2 != 0 {
+            return Err(EngineError::Io(io::Error::new(ErrorKind::InvalidInput,
+                                                      format!("{} mda size is not an even \
+                                                               number",
+                                                              path.display()))));
+        }
+
+        if mda_size < MIN_MDA_SIZE {
+            return Err(EngineError::Io(io::Error::new(ErrorKind::InvalidInput,
+                                                      format!("{} mda size is less than \
+                                                               minimum ({})",
+                                                              path.display(),
+                                                              *MIN_MDA_SIZE))));
+        }
+
+        let dev = try!(Device::from_str(&path.to_string_lossy()));
+
+        // map_err so we can improve the error message
+        let mut f = try!(OpenOptions::new()
+            .read(true)
+            .open(path)
+            .map_err(|_| {
+                io::Error::new(ErrorKind::PermissionDenied,
+                               format!("Could not open {}", path.display()))
+            }));
+
+        if !force {
```

Comment:
Did we agree? I'm afraid I don't recall so we may need to re-discuss :frowning_face: